pallets/runtime/tests/src/asset_metadata_test.rs

```diff
@@ -0,0 +1,427 @@
+        lock_status: if locked {
+            AssetMetadataLockStatus::Locked
+        } else {
+            AssetMetadataLockStatus::Unlocked
+        },
+    }
+}
+
+fn make_metadata_type(name: &str) -> (AssetMetadataName, AssetMetadataSpec) {
+    let spec = AssetMetadataSpec {
+        url: Some(b"http://example.com/test_specs".into()),
+        description: Some(format!("{} metadata type", name).as_bytes().into()),
+        type_def: Some(vec![]),
+    };
+    let name: AssetMetadataName = name.as_bytes().into();
+
+    (name, spec)
+}
+
+fn registe_metadata_type(owner: User, ticker: Option<Ticker>, name: &str) -> AssetMetadataKey {
+    let spec = AssetMetadataSpec {
+        url: Some(b"http://example.com/test_specs".into()),
+        description: Some(format!("{} metadata type", name).as_bytes().into()),
+        type_def: Some(vec![]),
+    };
+    let name: AssetMetadataName = name.as_bytes().into();
+
+    if let Some(ticker) = ticker {
+        // Register local metadata type with asset owner.
+        exec_ok!(Asset::register_asset_metadata_local_type(
+            owner.origin(),
+            ticker,
+            name.clone(),
+            spec,
+        ));
+
+        Asset::asset_metadata_local_name_to_key(ticker, name)
+            .map(AssetMetadataKey::from)
+            .expect("Failed to register metadata")
+    } else {
+        let root = frame_system::RawOrigin::Root;
+        // Register global metadata type with root.
+        assert_ok!(Asset::register_asset_metadata_global_type(
+            root.into(),
+            name.clone(),
+            spec,
+        ));
+
+        Asset::asset_metadata_global_name_to_key(name)
+            .map(AssetMetadataKey::from)
+            .expect("Failed to register metadata")
+    }
+}
+
+#[test]
+fn set_asset_metadata_local_type() {
+    ExtBuilder::default().build().execute_with(|| {
+        let owner = User::new(AccountKeyring::Dave);
+        let other = User::new(AccountKeyring::Alice);
+
+        // Create asset.
+        let (ticker, _) = create_token(owner);
+
+        let global_key = registe_metadata_type(owner, None, "TEST");
+        let local_key = registe_metadata_type(owner, Some(ticker), "TEST");
+
+        let value = AssetMetadataValue("cow".as_bytes().into());
+        let details = Some(make_metadata_value_details(None, false));
+
+        // Try adding global metadata value with user that doesn't have permissions for this asset.
+        exec_noop!(
+            Asset::set_asset_metadata(
+                other.origin(),
+                ticker,
+                global_key,
+                value.clone(),
+                details.clone()
+            ),
+            EAError::UnauthorizedAgent
+        );
+
+        // Try adding local metadata value with user that doesn't have permissions for this asset.
+        exec_noop!(
+            Asset::set_asset_metadata(
+                other.origin(),
+                ticker,
+                local_key,
```

Comment:
These are also always repeated so let's use that closure.

Reply:
no.